### DISCUSSION TITLE
Just wanted to say thanks!

### DISCUSSION THREAD

**Discussion by rwill128** [chrxh/alien#57] [General]
@chrxh I have been digging into this code, and having a blast editing it. 

I used to work on a program called Biogenesis on Sourceforge (http://biogenesis.sourceforge.net/) -- it is actually how I learned Java nd became a career programmer!

I have always wanted to take some of the concepts from that program and bring them into a CUDA and native-code environment, and your codebase is such an amazing and brilliant example of how to do that!

Really just wanted to say I think it's incredible what you've built here. What kind of education and professional experience do you have?

**Comment by chrxh** [comment]
Thank you, I'm glad to hear that. Unfortunately, there is no detailed technical documentation for it, just an overview for rough orientation https://alien-project.gitbook.io/docs/under-the-hood/architecture. It had taken a few iterations until the source code reached this state...

I was once working as a PostPoc on differential equations from continuum mechanics and their optimization. But now I'm working in industry as a C++ developer. 

Alien is so far only a rudimentary evolution simulator. For interesting simulations crucial elements are still missing. At the moment I'm experimenting with different ways to give the cell color/type a more specific meaning (e.g. to model multiple food sources and food chains), to simulate ecosystems in which different types of populations can coexist in the long term.

Biogenesis seems to be quite advanced in this regard. Maybe some ideas can be incorporated in a particle world like Alien?

**Reply by rwill128** [reply]
Biogenesis is a much less sophisticated program in some ways, mostly technically speaking but also performance-wise. For example I believe there is a lot of depth and potential to the assembly-like behavioral programs that you've given each cell in Alien.

However I think you are right, the possibilities for interesting evolutionary changes and complex evolved ecosystems are quite limited in Alien as it currently stands. I have some ideas on this and I am experimenting with some changes when I have time. We can discuss more in depth in the future as well.

**Reply by chrxh** [reply]
I've now added a few extensions that allow a bit more possibilities for evolution simulations: In particular a food chain color matrix and cell color transition rules can be configured:
![image](https://user-images.githubusercontent.com/73127001/195205669-edc63529-8682-440e-9277-5dc290cbcade.png)
Some small promising experiments I have already done (see https://youtu.be/hg89-r9Aj2w).

I think the assembler-like language for cell programs is useful for designing own creatures. However, it might be not necessarily so perfect for evolution processes by mutation and selection because the associated "fitness landscape" is very discontinuous.
That's why I'm currently working on incorporating neural networks.

On the new feature branch `features/3.4/NeuralNet` you will find the ongoing development to a new cell function that implements a layer of a neural network. A cell equipped with this function has 8 neurons and processes 8 input values from the token memory. The 8x8 weights and 8 bias values are stored in the cell memory. Optionally, a neural network cell can retrieve input values from "nearby" sensor cells in its own cluster (it does not have to be directly connected), and provide its output to nearby muscle and digestion cells.
In this way, a token carries neuronal activity information of one layer of an NN. By chaining multiple NN cells together, any number of layers can be be implemented in a cell cluster. Recurrent NNs are also possible when the token rotates in a loop. 

The neurons will be an addition to the existing cell functions, which will be kept as well.
